### PR TITLE
[Example] Add HGP-SL example for pytorch backend

### PR DESCRIPTION
**Pull request by lygztq**
## Description
* Add Hierarchical Graph Pooling with Structure Learning (HGP-SL) example for pytorch backend, paper link: https://arxiv.org/pdf/1911.05954.pdf

* Fix bug in sagpool example

* Fix path mistake for hardgat

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/hgp_sl/README.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+
+Example implementor
+----------------------
+This example was implemented by [Tianqi Zhang](https://github.com/lygztq) during his Applied Scientist Intern work at the AWS Shanghai AI Lab.
+
+
+The graph dataset used in this example 
+---------------------------------------
+The DGL's built-in LegacyTUDataset. This is a serial of graph kernel datasets for graph classification. We use 'DD', 'PROTEINS', 'NCI1', 'NCI109', 'Mutagenicity' and 'ENZYMES' in this HGP-SL implementation. All these datasets are randomly splited to train, validation and test set with ratio 0.8, 0.1 and 0.1.
```
Comment (hetong007) [inline examples/pytorch/hgp_sl/README.md:14]: Better add a link to the doc of LegacyTUDataset

File: examples/pytorch/hgp_sl/functions.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        scores = scores.float()
+        scores_max = _gspmm(gidx, "copy_rhs", "max", None, scores)[0]
+        scores = _gsddmm(gidx, "sub", scores, scores_max, "e", "v")
+
+        # find threshold for each node and perform ReLU(u-t(u)) operation.
+        tau, supp_size = _threshold_and_support_graph(gidx, scores, end_n_ids)
+        out = torch.clamp(_gsddmm(gidx, "sub", scores, tau, "e", "v"), min=0)
+        ctx.backward_cache = gidx
+        ctx.save_for_backward(supp_size, out)
+        torch.cuda.empty_cache()
```
Comment (hetong007) [inline examples/pytorch/hgp_sl/functions.py:115]: @yzh119 Is it the best practice?

File: examples/pytorch/hgp_sl/layers.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+from dgl.nn import AvgPooling, GraphConv, MaxPooling
+from dgl.ops import edge_softmax
+from torch import Tensor
+from torch.nn import Parameter
+
+from functions import edge_sparsemax
+from utils import get_batch_id, topk
+
+
+class WeightedGraphConv(GraphConv):
```
Comment (hetong007) [inline examples/pytorch/hgp_sl/layers.py:17]: @yzh119 Do we have such a GraphConv in our roadmap?

File: examples/pytorch/hgp_sl/networks.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+import torch
+from dgl.nn import AvgPooling, MaxPooling
+import torch.nn.functional as F
+import torch.nn
+from layers import HGPSLPool, WeightedGraphConv, ConvPoolReadout
+
+
+class SimpleModel(torch.nn.Module):
```
Comment (hetong007) [inline examples/pytorch/hgp_sl/networks.py:8]: 1. Is `SimpleModel` a special case of `Model`? If so, we may define this class by `Model`.
2. Better rename the classes to something more meaningful, e.g. `HGPSLModel`.

**Review by hetong007 — COMMENTED** [review]
Overall LGTM.

File: examples/pytorch/hgp_sl/README.md (hunk 1 above)
Comment (lygztq) [reply examples/pytorch/hgp_sl/README.md]: solved

File: examples/pytorch/hgp_sl/networks.py (hunk 4 above)
Comment (lygztq) [reply examples/pytorch/hgp_sl/networks.py]: The `SimpleModel` is a 3-layer (fixed, cannot change using input parameters) model which is the same as the author's implementation. We use the `Model` class which can set the number of layers from outside. I will remove the redundant `SimpleModel`

**Pull request merged**